File: frontend/app/src/main/java/com/example/nomo/viewmodel/AuthViewModel.java
package com.example.nomo.viewmodel;

import android.app.Application;
import android.content.Context;
import android.widget.Toast;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.nomo.api.AuthApi;
import com.example.nomo.model.AuthRequest;
import com.example.nomo.model.AuthResponse;
import com.example.nomo.utils.SharedPrefManager;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@HiltViewModel
public class AuthViewModel extends AndroidViewModel {
    private final MutableLiveData<Boolean> loginSuccess = new MutableLiveData<>();
    private final MutableLiveData<Boolean> registerSuccess = new MutableLiveData<>();

    private final AuthApi authApi;
    private final SharedPrefManager sharedPrefManager;

    @Inject
    public AuthViewModel(Application application,
                         AuthApi authApi,
                         SharedPrefManager sharedPrefManager) {
        super(application);
        this.authApi = authApi;
        this.sharedPrefManager = sharedPrefManager;
    }

    public LiveData<Boolean> getLoginSuccess() {
        return loginSuccess;
    }

    public LiveData<Boolean> getRegisterSuccess() {
        return registerSuccess;
    }

    public void registerUser(String nickname, String email, String password, Context context) {
        AuthRequest request = new AuthRequest(nickname, email, password);

        authApi.registerUser(request).enqueue(new Callback<AuthResponse>() {
            @Override
            public void onResponse(Call<AuthResponse> call, Response<AuthResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    sharedPrefManager.saveToken(response.body().getAccessToken());
                    sharedPrefManager.saveUserId(response.body().getUserId());
                    sharedPrefManager.saveUsername(nickname);
                    registerSuccess.postValue(true);
                } else {
                    registerSuccess.postValue(false);
                }
            }

            @Override
            public void onFailure(Call<AuthResponse> call, Throwable t) {
                registerSuccess.postValue(false);
            }
        });
    }

    public void loginUser(String nickname, String password, Context context) {
        AuthRequest request = new AuthRequest(nickname, "", password);

        authApi.loginUser(request).enqueue(new Callback<AuthResponse>() {
            @Override
            public void onResponse(Call<AuthResponse> call, Response<AuthResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    sharedPrefManager.saveToken(response.body().getAccessToken());
                    sharedPrefManager.saveUserId(response.body().getUserId());
                    sharedPrefManager.saveUsername(nickname);
                    loginSuccess.postValue(true);
                } else {
                    loginSuccess.postValue(false);
                }
            }

            @Override
            public void onFailure(Call<AuthResponse> call, Throwable t) {
                loginSuccess.postValue(false);
            }
        });
    }

    public void showError(String message, Context context) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
    }
}

File: frontend/app/src/main/java/com/example/nomo/viewmodel/AuthViewModel.kt
package com.example.nomo.viewmodel

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.nomo.managers.AuthManager
import com.example.nomo.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class AuthViewModel @Inject constructor(
    private val repository: AuthRepository
) : ViewModel() {

    sealed class AuthState {
        object Idle : AuthState()
        object Loading : AuthState()
        object Success : AuthState()
        data class Error(val message: String) : AuthState()
    }

    private val _authState = MutableStateFlow<AuthState>(AuthState.Idle)
    val authState: StateFlow<AuthState> = _authState

    fun resetState() {
        _authState.value = AuthState.Idle
    }

    fun registerUser(nickname: String, email: String, password: String, context: Context) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            try {
                val response = repository.registerUser(nickname, email, password)
                if (response.isSuccessful) {
                    val body = response.body()
                    if (body != null && body.token != null) {
                        AuthManager.saveToken(context, body.token)
                        _authState.value = AuthState.Success
                    } else {
                        _authState.value = AuthState.Error("Token missing in response")
                    }
                } else {
                    val error = response.errorBody()?.string() ?: "Unknown error"
                    _authState.value = AuthState.Error("Registration failed: $error")
                    AuthManager.clearToken(context)
                }
            } catch (e: Exception) {
                _authState.value = AuthState.Error("Network error: ${e.localizedMessage}")
            }
        }
    }

    fun loginUser(nickname: String, password: String, context: Context) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            try {
                val response = repository.loginUser(nickname, password)
                if (response.isSuccessful) {
                    response.body()?.token?.let { token ->
                        AuthManager.saveToken(context, token)
                        _authState.value = AuthState.Success
                    } ?: run {
                        _authState.value = AuthState.Error("Token missing from response on login")
                    }
                } else {
                    val error = response.errorBody()?.string() ?: "Unknown error"
                    _authState.value = AuthState.Error("Login error: ${error.take(200)}")
                }
            } catch (e: Exception) {
                _authState.value =
                    AuthState.Error("Network error: ${e.localizedMessage ?: "Unkown error"}")
            }
        }
    }

    fun showError(message: String) {
        _authState.value = AuthState.Error(message)
    }
}